Block & Unblock User Functionality

nabramow opened the pull request:
Closes #4165 
Closes #5962 

<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
**Please give clear steps for how the reviewer can best test this PR**

Please include any necessary dev environment, .env, etc. adjustments.

- Go to your profile => my connections
- Block a friend by clicking the ... and "block user"
- That user should appear in your "Blocked users" and be removed from your friends
- Now unblock them
- They should be removed from your blocked users and go back to your friends
- Go to a user's profile
- Click "Report or block" and check to block them, then confirm
- Go back to your connections, they should be in your blocked users
- Block someone that has discussions, messages, etc. with you
- Check the map search, discussions, host requests, messages etc. to make sure you can no longer see them anywhere
- Go to a message with that user, you should no longer be able to write them a message
- Check group chat with the blocked user, should say "Unknown user"

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on the web frontend)
If you need help with any of these, please ask :)
--->
**Web frontend checklist**
- [ x ] Formatted my code with `yarn format`
- [ x ] There are no warnings from `yarn lint --fix`
- [ x ] There are no console warnings when running the app
- [ x ] Added tests where relevant
- [ x ] All tests pass
- [ x ] Clicked around my changes running locally and it works
- [ x ] Checked Desktop, Mobile and Tablet screen sizes

<!---
Remember to request review from couchers-org/web, couchers-org/backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

nabramow committed:
Break out profile report flag and adjust style

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Finish dialog design and hook in block api

nabramow committed:
Do skeleton for blocked users list

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
WIP

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
WIP

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Get blocked user card properly rendering

nabramow committed:
Get blocking and unblocking working

nabramow committed:
Add BlockerdUserList tests

nabramow committed:
Clean up

nabramow committed:
Add dashboard news

nabramow committed:
Fix tests, types, mui styled imports

aapeliv commented:
If you block someone, then immediately go to their profile, it'll still show stuff. Maybe we need to purge the user from the query cache on block?

aapeliv commented:
Also the avatar in the unblock list links to the user's profile (which is no longer visible)

nabramow committed:
Remove blocked profile link and purge blocked profile from cache

aapeliv commented:
![Screenshot from 2025-06-01 09-37-59](https://github.com/user-attachments/assets/117ea76d-dd35-4d41-8ab7-1dd4d697746e)

- There's no snackbar/toast when you do block.
- If you go back to the user’s page it’s still there from cache (with error though), see above

jesseallhands commented on app/web/resources/locales/en.json:
```diff
@@ -69,14 +69,19 @@
     },
     "flag": {
+      "block_user": "Block {{username}}",
+      "block_user_explainer": "Messages and requests from this person will be blocked, and your profile, events and comments will not be visible to this person.",
```
```suggestion
      "block_user_explainer": "Messages and requests from this person will be blocked, and your profile, events and comments will not be visible to this person. They will not be informed that you blocked them.",
```

jesseallhands reviewed (changes requested):
The functionality works on the profile page, but it doesn't work in the messages or requests/hosting pages:

![image](https://github.com/user-attachments/assets/3dae3352-cecd-4517-801d-59d7d74c031a)

I think it should probably work wherever flags are -- but of course we'd only have the "Report or Block" text on the profile pages.

jesseallhands commented on app/web/features/connections/locales/en.json:
```diff
@@ -14,5 +22,11 @@
+  "unblock_user": "Unblock user",
+  "unblock_user_confirmation_dialog": {
+    "title": "Are you sure you want to unblock {{name}}?",
+    "message": "Unblocking will allow them to see your profile and send you requests. If you were previously friends, they will go back into your friends list.",
```
```suggestion
    "message": "Unblocking will allow them to see your profile and send you requests. If you were previously friends, they will go back into your friends list. They will not be notified.",
```

jesseallhands commented on app/web/features/connections/locales/en.json:
```diff
@@ -1,10 +1,18 @@
+  "block_user": "Block user",
+  "block_user_confirmation_dialog": {
+    "title": "Are you sure you want to block {{name}}?",
+    "message": "This will prevent you both from seeing each others profiles. You will not be able to send requests or messages.",
```
```suggestion
    "message": "Messages and requests from this person will be blocked, and your profile, events and comments will not be visible to this person. They will not be informed that you blocked them.",
```

jesseallhands reviewed (changes requested):
I added a couple more things since I forgot to check out the connections page and the functionality there (which looks great and works well!) One thing I'm not sure of: @aapeliv will have to confirm whether events and discussions/replies authored by the person will be hidden or not. If not, those two words can be left out of my suggestions. EDIT: Aapeli confirmed they cannot see their events or discussions.

When blocking someone, it seems like a report is sent to support sometimes even if the person doesn't choose to also report the person. Here's an example of what I mean (note that there is no "reason" given):

https://chatwoot.couchershq.org/app/accounts/2/conversations/3379

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Update app/web/features/connections/locales/en.json

Co-authored-by: Jesse Allhands <79384268+jesseallhands@users.noreply.github.com>

nabramow committed:
Update app/web/features/connections/locales/en.json

Co-authored-by: Jesse Allhands <79384268+jesseallhands@users.noreply.github.com>

nabramow commented:
> The functionality works on the profile page, but it doesn't work in the messages or requests/hosting pages:

> I think it should probably work wherever flags are -- but of course we'd only have the "Report or Block" text on the profile pages.

Hmm, this is not what was decided in the ticket. I also don't think that's usual behavior for blocking functionality. Let's leave it as is for now, we can discuss if we want to expand it later.

nabramow commented:
> When blocking someone, it seems like a report is sent to support sometimes even if the person doesn't choose to also report the person. Here's an example of what I mean (note that there is no "reason" given):

Oooo good catch, I fixed this now!

nabramow committed:
Fix redirect to user page and excess reporting

nabramow committed:
Merge remote-tracking branch 'origin/na/web/block' into na/web/block

nabramow committed:
Update app/web/resources/locales/en.json

Co-authored-by: Jesse Allhands <79384268+jesseallhands@users.noreply.github.com>

nabramow commented:
> * There's no snackbar/toast when you do block.

So this is trickier than it seems. Since after they block, the user profile is no longer visible to them, I can't render a snackbar there. I'd need to build some sort of higher up Toast at the app level, then pass it to there via React Context or something, but that's a bit of a feature in itself. I could redirect them to their connections page instead of the dashboard maybe?  Then they see the blocked user there.

> * If you go back to the user’s page it’s still there from cache (with error though), see above

This is fixed, I'm now routing them to the `NotFound` page if they go to a blocked profile.

aapeliv commented:
> > * If you go back to the user’s page it’s still there from cache (with error though), see above
> 
> This is fixed, I'm now routing them to the `NotFound` page if they go to a blocked profile.

I think it's better to leave as it was before. I don't understand react query well enough to know why it doesn't work and what corner of the cache the data is left to be in. But I think it's better not to have all that extra logic on the frontend just for this corner case (now I think this cause us to load blocked users every time you load the profile page).

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Merge remote-tracking branch 'origin/na/web/block' into na/web/block

nabramow committed:
Clear blocking user cacheand fix messages name and avatar

nabramow commented:
> > > * If you go back to the user’s page it’s still there from cache (with error though), see above
> > 
> > 
> > This is fixed, I'm now routing them to the `NotFound` page if they go to a blocked profile.
> 
> I think it's better to leave as it was before. I don't understand react query well enough to know why it doesn't work and what corner of the cache the data is left to be in. But I think it's better not to have all that extra logic on the frontend just for this corner case (now I think this cause us to load blocked users every time you load the profile page).

@aapeliv Okay I fixed the blocked user issue, but I think we need to filter out the messages from the blocked user in the MessagesList on the backend. I changed the username and avatar to "Unknown user", but it seems we're filtering out the user, but not their messages:

<img width="838" alt="Screenshot 2025-06-07 at 10 06 00 AM" src="https://github.com/user-attachments/assets/3c6c80c5-ed57-4f4e-bb8f-4d6e31e4b70a" />

jesseallhands reviewed (dismissed):
lgtm

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Merge branch 'develop' into na/web/block

nabramow committed:
Style adjustments for blocked users in messages

nabramow committed:
Merge branch 'develop' into na/web/block

aapeliv reviewed (approved):
LGTM
